Optimized marshaling of Go values into WAF values

Julio-Guerra opened the pull request:
The CPU profiler along with the new benchmark (agent side) allowed to spot every source of overhead.
    
The new implementation adds:
    
- requires at least go 1.12 which introduces `reflect.Value.MapRange()` which
      prevents using `reflect.Value.MapKeys()` that was creating a slice. Resulting
      in lots of overhead when large maps were traversed.
    
- strings are now limited to 4kB.
    
- remove every use of `[]byte(str)` that in fact involve a copy of `str` into a
      new slice of bytes.
    
- the WAF value constructors are rewritten in Go to avoid short back-and-forth C
      calls that were adding a lot of CGO overhead (mainly pointer checks).

The benchmarks now gives a maximum time of 1ms.

Commit:
waf: avoid string copies in go bindings when running a rule

Remove the string copy performed by `C.CString()` when creating a PWArgs string.
The PWArgs string constructor function will copy it into a new C char buffer. So
it is fine not to copy it during the time of the call.

This is a major source of performance improvement which allows to run much
faster, as most of the WAF inputs are strings.

Commit:
waf: remove unused log callback

We were not using it at all and involves many dependencies. By removing it, we
should not get the log-related function in the executable file, plus more lto
removals when possible.

Commit:
waf: optimized marshaler into waf values

The CPU profiler along with the new benchmark allowed to spot every source of
overhead.

The new implementation adds:

- requires at least go 1.12 which introduces `reflect.Value.MapRange()` which
  prevents using `reflect.Value.MapKeys()` that was creating a slice. Resulting
  in lots of overhead when large maps were traversed.

- strings are now limited to 4kB.

- remove every use of `[]byte(str)` that in fact involve a copy of `str` into a
  new slice of bytes.

- the WAF value constructors are rewritten in Go to avoid short back-and-forth C
  calls that were adding a lot of CGO overhead (mainly pointer checks).

The benchmarks now gives a maximum time of 1ms.

Commit:
ci: drop go1.11

Commit:
ci: remove -x to be less verbose

lloeki commented on waf/internal/bindings/waf.go:
```diff
@@ -194,16 +187,146 @@ func valueToWAFInput(v reflect.Value) (in *C.PWArgs, err error) {
+
+func (a *WAFArray) Index(i int) *WAFValue {
+	if C.uint64_t(i) >= a.nbEntries {
+		panic(errors.New("out of bounds access to WAFArray"))
```
Hmm, really not keen on those panics. Looks like you're beginning to use them a lot like exceptions instead of errors

Julio-Guerra replied:
Only because `reflect` uses them this way too and that binding accessors are only based on `reflect`. So I am using the same principle in case of unexpected behavior.

For example:
```
reflect.ValueOf([]int{1, 2, 3}).Index(33) // panic: out of bound access
```

So I panic too for this specific case.

lloeki commented on waf/types/types.go:
```diff
@@ -69,6 +71,7 @@ var (
 	_ error = ErrInvalidRule
 	_ error = ErrInvalidFlow
 	_ error = ErrNoRule
+	_ error = ErrOutOfMemory
```
Is that a new one defined in the libsqreen WAF spec?

Julio-Guerra replied:
No, custom for me with the Go reimplementation when `malloc()` or cgo allocation function (`CString()`, etc.) return nil.

lloeki commented:
LGTM, save for the `panic`.

Commit:
ci: drop alpine/386

No easy way to mix architectures on alpine. So dropping it for now.
The i386 target architecture is already covered with debian.

Commit:
waf/tests: fix mistakenly empty test string

gofuzz doesn't allow to enforce the string size. Rather use `rand.Read()` to
control the string size.

Commit:
ci: make alpine tests mandatory
